staging/src/k8s.io/mount-utils/mount_linux.go

```diff
@@ -550,6 +551,55 @@ func (mounter *SafeFormatAndMount) formatAndMountSensitive(source string, target
 	return nil
 }
 
+func (mounter *SafeFormatAndMount) format(fstype string, args []string) ([]byte, error) {
+	mounter.acquireFormatSem()
+
+	out := make(chan *formatResult, 1)
+	go mounter.execFormat(fstype, args, out)
```

Comment:
what happens if execFormat panics? if the goal is to release the semaphore after timeout and let the execFormat continue to block returning, it seems safer to put the semaphore release and timeout inside a goroutine, and leave the calls to mounter.Exec.Command completely alone and in the main goroutine:

```go
done := make(chan struct{})
defer close(done)

// get the semaphore
mounter.acquireFormatSem()
go func(){
  // release the semaphore when done or timeout is hit
  defer mounter.releaseFormatSem()
  // set up a timer that cleans up on exit
  timeout := time.NewTimer(mounter.formatTimeout)
  defer timeout.Stop()
  select {
    case <-done:
    case <-timeout.C:
  }
}()

return mounter.Exec.Command("mkfs."+fstype, args...).CombinedOutput()
```

Reply:
I'm also not sure the acquireFormatSem / releaseFormatSem indirection is helpful... would making the entire obtain/release flow conditional be clearer:

```go
	if m.formatSem != nil {
		// track when mount completes
		done := make(chan struct{})
		defer close(done)

		// block until a free slot opens up
		m.formatSem <- struct{}{}

		go func() {
			// release our slot
			defer func() { <-m.formatSem }()

			// set up a timer that cleans up on exit
			timeout := time.NewTimer(m.formatTimeout)
			defer timeout.Stop()

			// block until done or timed out
			select {
			case <-done:
			case <-timeout.C:
			}
		}()
	}

  return m.Exec.Command("mkfs."+fstype, args...).CombinedOutput()
```

Reply:
I was wondering how to use defers, but was struggling since my sem release logic was stuck in the center of the function. The done channel is brilliant.

The time.After() timer not cleaning up unless fired is very subtle (https://pkg.go.dev/time#After). Thanks for pointing that out.

Also, I haven't found a definitive decision on which channel type to use for a basic signal. After some thought, it seems like struct{} really is the only unary type.